src/callbacks/analysis.jl

```diff
@@ -0,0 +1,380 @@
+        end
+        println()
+      end
+
+      # Linf error
+      if :linf_error in analysis_errors
+        print(" Linf error:  ")
+        for v in eachvariable(equations)
+          @printf("  % 10.8e", linf_error[v])
+          analysis_callback.save_analysis && @printf(io, "  % 10.8e", linf_error[v])
+        end
+        println()
+      end
+
+      # Conservation errror
+      if :conservation_error in analysis_errors
+        @unpack analysis_callback = analysis_callback
+        state_integrals = integrate(u, semi)
+
+        print(" |∑U - ∑U₀|:  ")
+        for v in eachvariable(equations)
+          err = abs(state_integrals[v] - initial_state_integrals[v])
+          @printf("  % 10.8e", err)
+          analysis_callback.save_analysis && @printf(io, "  % 10.8e", err)
+        end
+        println()
+      end
+
+      # Residual (defined here as the vector maximum of the absolute values of the time derivatives)
+      if :residual in analysis_errors
+        print(" max(|Uₜ|):   ")
+        for v in eachvariable(equations)
+          # Calculate maximum absolute value of Uₜ
+          @views res = maximum(abs, view(du, v, ..))
+          @printf("  % 10.8e", res)
+          analysis_callback.save_analysis && @printf(io, "  % 10.8e", res)
+        end
+        println()
+      end
+
+      # L2/L∞ errors of the primitive variables
+      if :l2_error_primitive in analysis_errors || :linf_error_primitive in analysis_errors
+        l2_error_prim, linf_error_prim = calc_error_norms(cons2prim, semi, t)
+
+        print(" Variable:    ")
+        for v in eachvariable(equations)
+          @printf("   %-14s", varnames_prim(equations)[v])
+        end
+        println()
+
+        # L2 error
+        if :l2_error_primitive in analysis_errors
+          print(" L2 error prim.: ")
+          for v in eachvariable(equations)
+            @printf("%10.8e   ", l2_error_prim[v])
+            analysis_callback.save_analysis && @printf(io, "  % 10.8e", l2_error_prim[v])
+          end
+          println()
+        end
+
+        # L∞ error
+        if :linf_error_primitive in analysis_errors
+          print(" Linf error pri.:")
+          for v in eachvariable(equations)
+            @printf("%10.8e   ", linf_error_prim[v])
+            analysis_callback.save_analysis && @printf(io, "  % 10.8e", linf_error_prim[v])
+          end
+          println()
+        end
+      end
+
+
+      # additional
+      for quantity in analysis_integrals
+        res = analyze(quantity, du, u, t, semi)
+        @printf(" %-12s:", pretty_form_repl(quantity))
+        @printf("  % 10.8e", res)
+        analysis_callback.save_analysis && @printf(io, "  % 10.8e", res)
+        println()
+      end
+    end # GC.@preserve du_ode
+
+    println("-"^80)
+    println()
+
+    # Add line break and close analysis file if it was opened
+    if analysis_callback.save_analysis
+      println(io)
+      close(io)
+    end
+  end
+
+  u_modified!(integrator, false)
```

Comment:
This is necessary to tell DiffEq that we didn't modify the current solution value in our callback. Otherwise, [FSAL Runge-Kutta methods](https://en.wikipedia.org/wiki/Dormand%E2%80%93Prince_method) have to do additional computations which would waste ressources.

Reply:
OK. Maybe you can add a one-line comment in one of the next PRs to dev to explain this?

Reply:
Done in fb66b9c